internal/runtime/safecall.go

```diff
@@ -0,0 +1,69 @@
+package runtime
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+)
+
+// SafeCall safely calls a function using reflection. It handles potential
+// panics by recovering and returning an error. The function `fn` is expected
+// to be a function, and `args` are the arguments to pass to that function.
+// It returns the result of the function call (if any) and an error if one
+// occurred during the call or if a panic was recovered.
+func SafeCall(fn any, args ...any) (result any, err error) {
+	// Ensure fn is a function
+	fnValue := reflect.ValueOf(fn)
+	fnType := fnValue.Type()
+	if fnValue.Kind() != reflect.Func {
+		return nil, errors.New("fn is not a function")
+	}
+
+	// Defer a function to handle panics
+	defer func() {
+		if r := recover(); r != nil {
+			err = fmt.Errorf("recovered from panic: %v", r)
+		}
+	}()
+
+	// Convert args to reflect.Value slice
+	in := make([]reflect.Value, len(args))
+	for i, arg := range args {
+		if arg == nil {
+			if fnType.IsVariadic() && i >= fnType.NumIn()-1 {
+				// Get the type of the variadic argument's element (e.g., interface{})
+				variadicType := fnType.In(fnType.NumIn() - 1).Elem()
+				// Create a zero Value of that type
+				in[i] = reflect.Zero(variadicType)
+			} else {
+				// Regular arguments
+				in[i] = reflect.Zero(fnType.In(i))
+			}
+			continue
+		}
+		in[i] = reflect.ValueOf(arg)
+	}
+
+	// Call the function using reflection
+	out := fnValue.Call(in)
+
+	// Process the output
+	if len(out) == 0 {
+		return nil, nil
+	}
+
+	// If there's only one return value
+	result = out[0].Interface()
+	if len(out) == 1 {
+		return result, nil
+	}
+
+	// If there are two return values (assuming the second is an error)
+	if len(out) == 2 && out[1].Type().Implements(reflect.TypeOf((*error)(nil)).Elem()) {
+		err, _ = out[1].Interface().(error)
+		return result, err
+	}
+
+	// Handle other cases as needed
+	return result, nil
```

Comment:
I'm a bit surprised here

Maybe it's because you know the code better than me.

If a method that return more than 2 value, the code will return the first one when called with safe.

Then when there is more than 2, what you have to check for looking for an error is the last value returned, not the second.

So based on this, maybe your code should:
- raise an error when safe is used on a function that returns more than 2 values
- raise an error when the method returns 2 values, but the second is not an error

It might be impossible for now, because nothing returns this, but at least it would be covered

Reply:
This is a surprise for me too but the reason are simple : `go/template` cannot handle more 🚎 as we can see in the `text/template` package the safeCall method return only one or two ([Source Code](https://cs.opensource.google/go/go/+/master:src/text/template/funcs.go;drc=0c1627812406a76ac256a08bee985a9817372446;l=365))

but adding an hard stop to be sure no developer make this mistakes we can take the decision to add a security here, but that will be a double check from sprout and go/template

ex:
```go
var ErrTooManyReturnVals = errors.New("function returned more than 2 values")

// .....
	default:
		// More than 2 return values - this is not supported
		return nil, ErrTooManyReturnVals
```

WDYT ?

Reply:
I add more robustness to the safecall function to be sure that will never be "sprout" fault